src/Mvc/Mvc.Core/src/Filters/ResponseCacheFilterExecutor.cs

```diff
@@ -59,9 +59,9 @@ public void Execute(FilterContext context)
             throw new ArgumentNullException(nameof(context));
         }
 
-        if (!NoStore)
+        if (!(NoStore || _cacheProfile.Location == ResponseCacheLocation.None || _cacheLocation == ResponseCacheLocation.None))
```

Comment:
When `Location` is `ResponseCacheLocation.None`, no exception is thrown.

Reply:
Even with `Cache-Control: no-cache` configured, I think you still must set a duration. `no-cache` is a really bad name.

> Note that no-cache does not mean "don't cache". no-cache allows caches to store a response but requires them to revalidate it before reuse. If the sense of "don't cache" that you want is actually "don't store", then no-store is the directive to use.

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control#no-cache

Since clients are still expected to store the resource, it makes sense that clients expect the server to provide a duration for how long to do so. If you don't want to set a duration, why not use `NoStore` like the `InvalidOperationException` message suggests?

Reply:
> Even with `Cache-Control: no-cache` configured, I think you still must set a duration. `no-cache` is a really bad name (as is `ResponseCacheLocation.None` which maps to it, but those ships have sailed).
> 
> > Note that no-cache does not mean "don't cache". no-cache allows caches to store a response but requires them to revalidate it before reuse. If the sense of "don't cache" that you want is actually "don't store", then no-store is the directive to use.
> 
> https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control#no-cache
> 
> Since clients are still expected to store the resource, it makes sense that clients expect the server to provide a duration for how long to do so. If you don't want to set a duration, why not use `NoStore` like the `InvalidOperationException` message suggests?

I know that no-cache means that the client can cache, but must make a request to the server. I mean in some cases it's impossible to predict when the data will change, it's not a good thing to let the client use the cache directly without making any requests to the server. When no-cache is set, the client must make a request to the server, and if the data has not changed, the server can return an HTTP 304 status code to save traffic.

Reply:
If the `no-cache` behavior is the behavior you want, great! This means that clients will store the resource to potentially reuse later if the server indicates the resource has not changed. This means you must set a duration for the client to store the resource though.

Reply:
I think `no-cache` already indicates that **in any case** the client must make a request to the server, and setting `max-age` is redundant.

Reply:
I think `Cache-Control: no-cache` is enough to express the server's requirements, no need to provide `max-age`.

Reply:
@Tratcher I'm not sure what the spec has to say about this, but I'm hesitant to change the behavior considering the "// Duration MUST be set (either in the cache profile or in this filter) unless NoStore is true." comment.

I really don't see the problem with just responding with `Cache-Control: no-cache,max-age=31536000` in any case. I doubt many clients are going to store resource for more than a year anyway.

Reply:
Although I think `no-cache` is sufficient, it's worth mentioning that in the code behind I changed the behavior to ignore `Duration` when `Location` is set to `ResponseCacheLocation.None`, but if you want to support `no-cache` at the same time And `no-cache,max-age=xxx`, the code behind also needs to be changed.